rust/src/graph.rs

```diff
@@ -0,0 +1,435 @@
+        for file in files.as_ref() {
+            graph_contents.push(std::fs::read(file).map_err(Into::<Error>::into)?);
+        }
+        let graph_builder_array: Vec<&[u8]> = graph_contents.iter().map(|s| s.as_ref()).collect();
+        let graph_handle =
+            syscall::load(graph_builder_array.as_slice(), self.encoding, self.target)?;
+        Ok(Graph {
+            build_info: self,
+            graph_handle,
+        })
+    }
+}
+
+/// An execution graph for performing inference (i.e., a model), which can create instances of [`GraphExecutionContext`].
+///
+/// ### Example
+/// ```rust
+/// use wasi_nn::{GraphBuilder, GraphEncoding, ExecutionTarget};
+///
+/// let model_file = "./test.tflite";
+/// // create a graph using `GraphBuilder`
+/// let graph = GraphBuilder::new(GraphEncoding::TensorflowLite, ExecutionTarget::CPU).build_from_files([model_file])?;
+/// // create an execution context using this graph
+/// let mut graph_exec_ctx = graph.init_execution_context()?;
+/// // set input tensors
+/// // ......
+///
+/// // compute the inference on the given inputs
+/// graph_exec_ctx.compute()?;
+///
+/// // get output tensors and do post-processing
+/// // ......
+/// ```
+pub struct Graph {
+    build_info: GraphBuilder,
+    graph_handle: syscall::GraphHandle,
+}
+
+impl Graph {
+    /// Get the graph encoding.
+    #[inline(always)]
+    pub fn encoding(&self) -> GraphEncoding {
+        self.build_info.encoding
+    }
+
+    /// Get the graph execution target.
+    #[inline(always)]
+    pub fn execution_target(&self) -> ExecutionTarget {
+        self.build_info.target
+    }
+
+    /// Get the graph id.
+    #[inline(always)]
+    pub fn graph_id(&self) -> syscall::GraphHandle {
+        self.graph_handle
+    }
+
+    /// Use this graph to create a new instances of [`GraphExecutionContext`].
+    #[inline(always)]
+    pub fn init_execution_context(&self) -> Result<GraphExecutionContext, Error> {
+        let ctx_handle = syscall::init_execution_context(self.graph_handle)?;
+        Ok(GraphExecutionContext {
+            graph: self,
+            ctx_handle,
+        })
+    }
+}
+
+/// Bind a [`Graph`] to the input and output for an inference.
+pub struct GraphExecutionContext<'a> {
+    graph: &'a Graph,
+    ctx_handle: syscall::GraphExecutionContextHandle,
+}
+
+impl<'a> GraphExecutionContext<'a> {
+    /// Get the [`Graph`] instance for this [`GraphExecutionContext`] instance.
+    #[inline(always)]
+    pub fn graph(&self) -> &Graph {
+        self.graph
+    }
+
+    /// Get the execution context id.
+    #[inline(always)]
+    pub fn context_id(&self) -> syscall::GraphExecutionContextHandle {
+        self.ctx_handle
+    }
+
+    /// Set input uses the `data`, not only [u8], but also [f32], [i32], etc.
+    pub fn set_input<T: Sized>(
+        &mut self,
+        index: usize,
+        tensor_type: TensorType,
+        dimensions: &[usize],
+        data: impl AsRef<[T]>,
```

Comment:
I kind of expected that this would be a `Tensor` struct. Is there a reason to prefer to pass the parts of a `Tensor` separately here and in `set_output`?

Reply:
In the `get_output`:
```rust
pub unsafe fn get_output(
    context: GraphExecutionContext,
    index: u32,
    out_buffer: *mut u8,
    out_buffer_max_size: BufferSize,
) -> Result<BufferSize, NnErrno>
```
It only needs an out buffer and does not need tensor type and tensor dimensions, so I use `out_buf: &mut [u8]` as a parameter instead of an entire `Tensor`.

In the `set_input`, It needs a `Tensor` as input. There are two reasons to use the parts of `Tensor` separately as parameters:
1. The user's `Tensor` may have different structures, if we use `wasi_nn::Tensor` as the `set_input` parameter, users should convert `user::Tensor` to `wasi_nn::Tensor` before calling.
2. It's difficult to define a `wasi_nn::Tensor` as the input parameter to support multiple types of `Tensor' 's data. The users may use owned data as input, such as `Vec<f32>`, or use referenced data as input to reuse the buffer, such as `&Vec<f32>`, `&[u8]`.

That's all, thanks!

Reply:
Yeah, I see what you mean about `set_output`. We could probably do better (like, more ergonomic) if we had `get_input_types` and `get_output_types` in the specification as mentioned [here](https://github.com/bytecodealliance/wasi-nn/issues/68#issuecomment-1295485435) but that does not need to gate this PR. We can revisit this once those new functions are added.